google/cloud/bigtable/read_rows_query.py

```diff
@@ -197,18 +241,113 @@ def add_range(
 
         Returns:
             - a list of queries that represent a sharded version of the original
               query (if possible)
+        Raises:
+            - AttributeError if the query contains a limit
         """
-        raise NotImplementedError
+        if self.limit is not None:
+            raise AttributeError("Cannot shard query with a limit")
+        if len(self.row_keys) == 0 and len(self.row_ranges) == 0:
+            # empty query represents full scan
+            # ensure that we have at least one key or range
+            full_scan_query = ReadRowsQuery(
+                row_ranges=RowRange(), row_filter=self.filter
+            )
+            return full_scan_query.shard(shard_keys)
+
+        sharded_queries: dict[int, ReadRowsQuery] = defaultdict(
+            lambda: ReadRowsQuery(row_filter=self.filter)
+        )
+        # the split_points divde our key space into segments
+        # each split_point defines last key that belongs to a segment
+        # our goal is to break up the query into subqueries that each operate in a single segment
+        split_points = [sample[0] for sample in shard_keys if sample[0]]
+
+        # 1. handle row_keys
+        # use binary search to find the segment that each key belongs to
+        for this_key in list(self.row_keys):
+            # bisect_left: in case of exact match, pick left side (i.e. keys are inclusive ends)
+            segment_index = bisect_left(split_points, this_key)
+            sharded_queries[segment_index].add_key(this_key)
+
+        # 2. handle row_ranges
+        for this_range in self.row_ranges:
+            # 2a. find the index of the segment the start key belongs to
+            if this_range.start is None:
+                # if range is open on the left, include first segment
+                start_segment = 0
+            else:
+                # use binary search to find the segment the start key belongs to
+                # bisect method determines how we break ties when the start key matches a split point
+                # if inclusive, bisect_left to the left segment, otherwise bisect_right
+                bisect = bisect_left if this_range.start.is_inclusive else bisect_right
+                start_segment = bisect(split_points, this_range.start.key)
+
+            # 2b. find the index of the segment the end key belongs to
+            if this_range.end is None:
+                # if range is open on the right, include final segment
+                end_segment = len(split_points)
+            else:
+                # use binary search to find the segment the end key belongs to.
+                # optimization: remove keys up to start_segment from searched list,
+                # then add start_segment back to get the correct index
+                cropped_pts = split_points[start_segment:]
+                end_segment = (
+                    bisect_left(cropped_pts, this_range.end.key) + start_segment
```

Comment:
Not for the end segment, since the `split_points` mark the inclusive ends of each segment. Whether the range includes or excludes the split point, the results will be the same (left side)

Do you think I should add a comment to call this out?